### PR TITLE
assess Francois' issues

### LINKED ISSUE #65: The FEC channel carries source symbols
In section `2.1.  Separate channels, separate entities`

This is an echo from a previous issue  that was not really addressed or "resolved". I will rephrase the problem from scratch and let me know if you agree with me or not. :-)

We describe the FEC channel (maybe should we call it the Coding channel as repair symbols could also be sent a-posteriori) and say that it carries repair symbols while the CC channel carries source packets.

What do you think of rephrasing it like the following: "The CC channel carries network packets and the coding channel carries source symbols and repair symbols" ?

source symbols are an abstract view on the application (and sometimes control) data that we want to protect. For example, several source symbols could be present in a single network packet, or several network packets could form one single source symbol. This is why network packets can be very different than source symbols. This also goes in our vision that CC and NC are separated channels and separated entities: they do not even share the concept of packets. 

What do you think ?

### LINKED ISSUE #66: Section 2.3 and Figure 4 not totally clear
I am not totally sure of the meaning of section 2.3 and Figure 4. More specifically: are we saying that multiple data streams transiting through one single protocol instance are out of the scope of the document ? Do we consider only one single stream on one single protocol instance ? If this is the case, does that mean that we do not consider QUIC's use-case ?  What makes me think all of this is the following sentence:

> The document considers one application layer stream as input packets above a combination of FEC and transport layers.

### LINKED ISSUE #67: avoid talking about linear systems
In section 2.4 :

>  In particular, if the codec does
>  not enable a subset of the linear system to be inverted, the
>  receiver would have to wait for a certain minimum amount of repair
>  packets before it can recover all the source packets.

Why are we talking about a linear system here ? To be more general on the recovery algorithms that are used, maybe should we use the term "block or window" instead of "linear system"

What do you think ?

### LINKED ISSUE #68: If the size of the window/block increase, the chances of recovering decrease
In section 2.4, we say the following:

> If the size
> of the encoding window (for Sliding Window Coding) and the size of
> the blocks (for Block Coding) are large, the chances of recovering
> the erased symbols would increase.  

Either I misunderstood the statement or the sentence is not true. In my opinion, with an equal number of sent repair symbols, if the window/block size is larger, the recovery probability is lower as they are likely more lost source symbols in the window/blocks (with a fixed number of repair symbols).

### LINKED ISSUE #69: Section 3, description of "fully reliable"
In section 3, we can read the following:

> This approach requires that the transport protocol does not implement
>  a fully reliable data transfer service (e.g., based on lost packet
>  retransmission).

I would emphasize on en in-order delivery which is the real problem for FEC here, as a reliable messaging system without in-order delivery would work.

I would thus replace the parenthesis by the following:

> (i.e. delivering the data in-order to the upper layer)

### LINKED ISSUE #70: Section 4 Figure 6
It is a bit complicated for me to understand correctly the details of Figure 6 :  

```
    | source  | sending                    ^ source
    | packets | rate                       | packets
    v         v                            |
   +------------+                      +------------+
   | Transport  |                      | Transport  |
   |            |                      |            |
   | +---+ +--+ |             signaling| +---+ +--+ |
   | |FEC| |CC| |              repaired| |FEC| |CC| |
   | +---+ +--+ |source         symbols| +---+ +--+ |
   |            |and/or             <==|            |
   |            |repair         network|            |
   |            |packets    information|            |
   +------------+ ==>               <==+------------+

       SENDER                              RECEIVER

                       Figure 6: FEC in the transport
```

In my opinion, the app does not sublit packets nor a sending rate to the transport layer, so the two arrows above the sender seem a bit weird for me. The app generally submits data, sometimes messages and maybe a minimal transmission rate but it is the transport layer that packetizes the data and sends packets.

I guess that the arriws between the sender and the receiver are the wire or the layers below transport. Thus I would replace "source and/or repair packets" by "packets containing source and/or repair symbols"

Did I misunderstand the Figure ?

### LINKED ISSUE #71: "transport and FEC mechanism could be conjointly designed" for partial reliability
In section 4.6: 

> The application may require partial reliability.  In this case, the
> transport and FEC mechanisms could be conjointly designed.  As one
> example, the reliability offered by FEC may be sufficient, with no
>  retransmission required.  This depends on application needs and the
>  trade-off between latency and loss.  Partial reliability impacts the
>  type of FEC and type of codec that can be used, such as discussed in
>  Section 2.4.

With coding, you can recover packets a-priori and a-posteriori, so I do not see any conjoint design between coding and the transport layer: if you replace the retransission mechanism by coding, you are more general as you can also recover packets a-priori and coding can offer full reliability if needed or partial reliability if the full reliability is not necessary. So I do not see any need of a conjoint design here: simply replace the retransmission-based reliability mechanism by the coding-based one. I do not really understand what we are trying to say here talking about a conjoint design.

What do you think ?

### PR DESCRIPTION
**Pull request by NicoKos**

### CODE REVIEW

**Comment by NicoKos** [comment]
I think this PR solves most of the issues that have been raised (and will close them if we merge this PR). 
What do you think @francoismichel @mwelzl @elochin ?

**Comment by mwelzl** [comment]
I'm ok with whatever all others decide  :)

File: draft-irtf-nwcrg-coding-and-congestion-08.xml [hunk 1]
```diff
@@ -247,8 +266,8 @@
...
-			<t>Partial reliability: The receiver cannot deliver source packets that could not have been decoded to the upper layer. If the size of the encoding window (for Sliding Window Coding) and the size of the blocks (for Block Coding) are large, the chances of recovering the erased symbols would increase. However, this would impact on memory requirements and the cost of encoding and decoding processes.</t>
+			<t>Full reliability: The receiver may hold symbols until the decoding of source symbols is possible. In particular, if the codec does not enable a subset of the system to be inverted, the receiver would have to wait for a certain minimum amount of repair packets before it can recover all the source symbols.</t>
+			<t>Partial reliability: The receiver cannot deliver source symbols that could not have been decoded to the upper layer. Increasing the amount of repair symbols would increase (1) the size of the encoding window (for Sliding Window Coding) or of the blocks (for Block Coding) and (2) the chances of recovering the erased symbols. However, this would impact on memory requirements and the cost of encoding and decoding processes.</t>
```
Comment (francoismichel) [inline draft-irtf-nwcrg-coding-and-congestion-08.xml]: I do not think that we can say that increasing the number of repair symbols also increases the window size, when we talk about the coding window size, we often consider the number of source packets that are present in the window. So in my opinion, adding more repair symbols does not increase the window size as it generally keeps the number of source symbols unchanged.

**Review by francoismichel — COMMENTED** [review]
Thank you for this great PR. Only one comment, otherwise this is good for me !

File: draft-irtf-nwcrg-coding-and-congestion-08.xml (hunk 1 above)
Comment (NicoKos) [reply draft-irtf-nwcrg-coding-and-congestion-08.xml]: Does the following commit https://github.com/irtf-nwcrg/draft-irtf-nwcrg-coding-and-congestion-in-transport/pull/72/commits/1c70162f201aacbd4e6e2fe2fe51600fea59bed0 makes things better?

**Pull request merged**